design/client-initiated-backchannel-authentication-flow.md

```diff
@@ -0,0 +1,494 @@
+
+### Default Implementation of AuthN/AuthZ by AD part of Outside CIBA Protocol
+
+There are several ways of AuthN/AuthZ by AD (e.g. push notification to an android device, push notification to an iOS device, send SMS message to an android device, send SMS message to an iOS device, make a telephone call, etc.). Each of such the way uses different device and technology so that keycloak cannot cover all of such ways. 
+
+Considering this point, this default implementation delegates AuthN/AuthZ by AD to other entity called here "Decoupled Auth Server" outside keycloak. It assumes that this Decoupled Auth Server is under control by an administrator of keycloak.
+
+```
+---------------------------------+                    +-----------+
+| keycloak                       |                    | Decoupled |
+|  +--------------------------+  |                    | Auth      |
+|  | Outside CIBA             |  |                    | Server    |
+|  |  +--------------------+  |  | (i) POST           |           |
+|  |  | AuthN/AuthZ by AD  | --[Decoupled Auth ID]--> |           |
+|  |  |                    | <-[Decoupled Auth ID]--- |           |
+|  |  +--------------------+  |  |          (ii) POST |           |
+|  +------------|-------------+  |                    |           | 
+|         (iii) |                |                    |           |
+|         [Auth Result ID]       |                    |           |
+|               V                |                    |           |
+|     +--------------------+     |                    |           |
+|     | Auth Result Cache  |     |                    |           |
+|     |                    |     |                    |           |
+|     +--------------------+     |                    |           |
++--------------------------------+                    +-----------+
+```
+
+(i)   : keycloak sends to Decoupled Auth Server an AuthN/AuthZ by AD request with "Decoupled Auth ID" bound with "Auth Result ID" bound with "auth_req_id" issued to CD.
+
+(ii)  : After completion of AuthN/AuthZ by AD, Decoupled Auth Server sends to keycloak the results of AuthN/AuthZ by AD with "Decoupled Auth ID".
+
+(iii) : After receiving the results of AuthN/AuthZ by AD, AuthN/AuthZ by AD part pushes this result to Auth Result Cache with "Auth Result ID" bound with "auth_req_id" issued to CD.
+
+AuthN/AuthZ by AD part's necessary tasks is as follows :
+
+1. Requests an AD to authenticate and get consents from an end-user
+2. Executes all the processes that need to be executed after the end-user logged-in like creating User Session and Authenticated Client Session that are needed to generate tokens in Token Req/Res part.
+3. Informing Token Req/Res part of 1's result.
+
+1 is highly dependent on actual use-cases while 2 and 3 seems to be use-case independent. Therefore, The abstract base class implementing 2 and 3 is provided and the class extending this abstract base class is also provided.
```

Comment:
Would #1 be an SPI?

Can we use security event tokens for this communication?

Reply:
Yes, #1 would be an SPI.

Is "security event tokens" you mentioned [RFC 8417 Security Event Token (SET)](https://tools.ietf.org/html/rfc8417) ? I've not yet read it so that I'd like to investigate into it.

Reply:
Yes, that is the spec. In fact, we should have been using this for other types of events we use (adapter, logout, etc).

Reply:
I see, I would like to use Security Event Token(SET). AFAIK, the current keycloak does not yet support SET so that we need to support it at first before implementing CIBA. Is that correct?

Reply:
It is a suggestion. We don't necessarily need to consider everything in the spec but provide some standard format for the events we send to and get from AD.

If you think it is too much, that is fine. But it would be much better than come up with our own format.

Reply:
@pedroigor Thanks. I'd like to decide whether SET is used as the formal JWT profile or not. But it is beneficial for us to implement SET because it can be used not only CIBA but also pushing logout security events to RPs as you mentioned.